### PR TITLE
Add special continuation rules to control behaviour at the route/stop level

### PR DESCRIPTION
**Pull request by npaun**
This PR adds a flexible rules system that allows users use knowledge about a particular feed to specify whether or not in-seat transfers should be allowed at certain routes and stops. This syntax is documented in the `README.md` and the overall design is inspired by OneBusAway rules.

In this initial PR, only the `modify` operation (change transfer_type) is supported. `remove` (delete any predicted transfers matching criteria) is easy to add if we ever need it. `create_block` (automatically group trips into blocks and then predict transfers from them) is a bit more work and could be added later.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,8 +34,62 @@ Riders probably won't be able to, or want to, to stay on board if:
...
+
+- _All_ selectors will match any trip-to-trip transfers.
+    - Example: `{"all": true}`
+- _Through_ selectors will match either the `from_trip` or the `to_trip`.
+    - Example: `{"through": {"route": "1", "stop": "Terminus Longueuil"}}`
+    - You can specify a `route` (route short name), a `stop` (stop name), or both.
+- _From_ selectors will match on the `from_trip`.
+    - Example: `{"from": {"route": "20", "last_stop": "Osachoff / White"}}`
+    - You can specify a `route`, a `last_stop`, or both.
+- _To_ selectors will match on the `to_trip`.
```
Comment (jsteelz) [inline README.md:63]: We should have a point like this

```
    - You can specify a `route`, a `last_stop`, or both.
``` for the To section

File: README.md [hunk 2]
```diff
@@ -34,8 +34,62 @@ Riders probably won't be able to, or want to, to stay on board if:
...
+- `op`: The action to take on matching predicted continuations.
+    - `modify`: Change the `transfer_type`.
+    - More operations are planned in the future.
+- `transfer_type`: Only for `modify` operations. The new `transfer_type` to assign,
+    - `4`: in-seat transfer
+    - `5`: vehicle continuation only
+
+The following _selectors_ are supported:
+
+- _All_ selectors will match any trip-to-trip transfers.
```
Comment (jsteelz) [inline README.md:55]: It should be stated explicitly that all and through selectors may not be combined with other selectors

File: README.md [hunk 3]
```diff
@@ -34,8 +34,62 @@ Riders probably won't be able to, or want to, to stay on board if:
...
+    - You can specify a `route` (route short name), a `stop` (stop name), or both.
+- _From_ selectors will match on the `from_trip`.
+    - Example: `{"from": {"route": "20", "last_stop": "Osachoff / White"}}`
+    - You can specify a `route`, a `last_stop`, or both.
+- _To_ selectors will match on the `to_trip`.
+    - Example:  `{"to": {"route": "5T", "first_stop": "Stewart Creek"}}`
+- _From_ and _to_ selectors can be combined to match a particular continuation between two trips].
+    - Example: `{"from": {"route": "124", "last_stop": "3rd / Pine"}, "to": {"route": "26", "first_stop": "3rd / Pine"}}`
+
+When a transfer matches multiple rules, the last rule wins. For transfers predicted from blocks, special continuation rules override all heuristics except `max_wait_time`. Rules never apply to predefined transfers specified using `transfers.txt`
```
Comment (jsteelz) [inline README.md:68]:
```suggestion
When a transfer matches multiple rules, the last rule wins. For transfers predicted from blocks, special continuation rules override all heuristics except `max_wait_time`. Rules never apply to predefined transfers specified using `transfers.txt`.
```

File: blocks_to_transfers/classify_transfers.py [hunk 4]
```diff
@@ -98,3 +114,89 @@ def get_shape_ptr(shape_match, trip):
...
+
+@dataclass
+class StandardSelector:
+    route: Optional[str] = None
+    stop: Optional[str] = None
+
+    def applies(self, trip, stop_to_check):
+        if self.route is None and self.stop is None:
+            # Has no criteria; probably a bug
+            return False
```
Comment (jsteelz) [inline blocks_to_transfers/classify_transfers.py:156]: We can throw and explicitly forbid this if we're feeling spicy

Comment (npaun) [reply blocks_to_transfers/classify_transfers.py]: I bumped this, and a few other problems up to fatal errors.

**Review by jsteelz — APPROVED** [review]

**Pull request merged**